docs/architecture/adr-053-state-sync-prototype.md

```diff
@@ -192,14 +191,96 @@ Taking consistent snapshots of IAVL trees is greatly simplified by them being ve
 
 Snapshots must also be garbage collected after some configurable time, e.g. by keeping the latest `n` snapshots.
 
-## Open Questions
+## Experimental Prototype
 
-* Is it possible to reconstruct an identical IAVL tree given separate subtrees in an appropriate order, or is more data needed about the branch structure?
+An experimental but functional state sync prototype is available in the `erik/statesync-prototype` branches of the Tendermint, IAVL, Cosmos SDK, and Gaia repositories. To fetch the necessary branches:
 
-* Should we punish nodes that provide invalid snapshots?
+```sh
+$ mkdir statesync
+$ cd statesync
+$ git clone git@github.com:tendermint/tendermint -b erik/statesync-prototype
+$ git clone git@github.com:tendermint/iavl -b erik/statesync-prototype
+$ git clone git@github.com:cosmos/cosmos-sdk -b erik/statesync-prototype
+$ git clone git@github.com:cosmos/gaia -b erik/statesync-prototype
+```
+
+To spin up three nodes of a four-node testnet:
+
+```sh
+$ cd gaia
+$ ./tools/start.sh
+```
+
+Wait for the first snapshot to be taken at height 3, then (in a separate terminal) start the fourth node with state sync enabled:
+
+```sh
+$ ./tools/sync.sh
+```
+
+To stop the testnet, run:
+
+```sh
+$ ./tools/stop.sh
+```
+
+## Open Questions
 
 * Is it OK for state-synced nodes to not have historical blocks nor historical IAVL versions?
 
+* Do we need incremental chunk verification for first version?
+
+* Should we call these snapshots? IAVL already has a different concept called snapshots.
+
+* Should the snapshot ABCI interface be a separate optional ABCI service, or mandatory?
+
+* How can we make sure `ListSnapshots` data is valid? An adversary can provide a large number of invalid snapshots and make the node stuck in discovery mode. Perhaps collect snapshot list from many peers and pick ones that are available on a majority.
+
+* Should we punish nodes that provide invalid snapshots? How?
+
+* Should we store snapshot and chunk metadata in a database? Can we use the database for chunks?
+
+* How large chunks can we send without getting memory issues? Can we stream the chunk contents?
```

Comment:
We will probably need to cap chunk size

Reply:
Yes, it's capped at 64 MB. But if we're downloading e.g. 8 chunks in parallel, that's 0.5 GB memory usage. And I think this is initially limited by the gRPC response buffer size, which is probably on the order of 100 MB.

Reply:
> gRPC response buffer

at what stage we're using gRPC?

Reply:
Sorry, by habit I just automatically think gRPC whenever I'm sending Protobuf requests. Does the p2p layer buffer entire messages in memory, or can it also do streaming processing?

Reply:
the answer is somewhere in this file https://github.com/tendermint/tendermint/blob/master/p2p/conn/connection.go#L1